rllib/core/rl_module/rl_module.py

```diff
@@ -378,35 +403,89 @@ def get_state(self) -> Mapping[str, Any]:
+            - the ray commit hash used
+            - the date and time of the checkpoint was created
 
-        Args:
-            state: The serialized state of the module.
+        """
+        gmt_time = datetime.datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S GMT")
+        metadata = {}
+        metadata["module_class"] = serialize_type(self.__class__)
+        metadata["module_config"] = self.config.to_dict()
+        metadata["ray_version"] = ray.__version__
+        metadata["ray_commit_hash"] = ray.__commit__
+        metadata["checkpoint_date_time"] = gmt_time
+        metadata["module_state_path"] = str(module_state_path)
+
+        framework = ""
+        if str(module_state_path).endswith(".pt"):
+            framework = "torch"
+        elif str(module_state_path).endswith("module_state"):
+            framework = "tf"
+        else:
+            raise ValueError(
+                "Unknown format for module state path must be either "
+                "'module_state' for tf or .pt for torch: "
+                f"{module_state_path}"
+            )
+        metadata["framework"] = framework
+        metadata_path = checkpoint_dir / "rl_module_metadata.json"
+        with open(metadata_path, "w") as f:
+            json.dump(metadata, f)
 
-        NOTE: this state is typically obtained from `serialize()`.
+    @classmethod
+    def _from_metadata_file(cls, metadata_dir_path: pathlib.Path) -> "RLModule":
+        """Constructs a module from the metadata.
 
-        NOTE: This method needs to be implemented in order to support
-            checkpointing and fault tolerance.
+        Args:
+            metadata: The metadata of the module.
 
         Returns:
-            A deserialized RLModule.
+            The module.
         """
-        module_class = state["class"]
-        config = RLModuleConfig.from_dict(state["config"])
-        module = module_class(config)
-        module.set_state(state["state"])
+        if not metadata_dir_path.exists():
+            raise ValueError("The metadata directory was not found.")
+        if not metadata_dir_path.is_absolute():
+            raise ValueError("The metadata directory path must be absolute.")
+        if not metadata_dir_path.is_dir():
+            raise ValueError("The metadata directory path must be a directory path.")
+        metadata_path = metadata_dir_path / "rl_module_metadata.json"
+        if not metadata_path.exists():
+            raise ValueError(
+                "While constructing the module from the metadata, the "
+                f"metadata file was not found at {str(metadata_path)}"
+            )
+        with open(metadata_path, "r") as f:
+            metadata = json.load(f)
+        module_class = deserialize_type(metadata["module_class"])
+        module_config = RLModuleConfig.from_dict(metadata["module_config"])
+        module = module_class(module_config)
         return module
 
+    def save_to_checkpoint(self, checkpoint_dir_path: str) -> None:
```

Comment:
I think both save_to_checkpoint and load_from_checkpoint should be abstract from RLModule. There are certain things that are common for all framework specific stuff. For example you need to do

```
        path = pathlib.Path(checkpoint_dir_path)
        if not path.exists():
            raise ValueError(
                "While loading from checkpoint there was no directory"
                " found at {}".format(checkpoint_dir_path)
            )
        if not path.is_absolute():
            raise ValueError("dir_path must be an absolute path.")
        if not path.is_dir():
            raise ValueError(
                "While loading from checkpoint the checkpoint_dir_path "
                "provided was not a directory."
            )

        module = cls._from_metadata_file(path)
        module_state_dir = path / "module_state"
        if not module_state_dir.exists():
            raise ValueError(
                "While loading from checkpoint there was no 'module_state'"
                " directory found at {}".format(checkpoint_dir_path)
            )
        module_state_dir.mkdir(parents=True, exist_ok=True)
 ```
 for both torch, tf or even jax. So put them in this base class implementation and define a new abstraction for framework-specific classes to implement. say set_state_from_file(...), and save_state_to_file(...)

Reply:
done